rfc/0014-attribute-based-valuetypes/README.md

```diff
@@ -42,47 +37,61 @@ valuetype CorrelationCoefficient oftype decimal {
 }
 ```
 
-This syntax is smart in that you don't have to list and name an attribute but rather rely on an implicit 'value' attribute. Still, I propose to make that attribute explicit. New syntax would be:
+This syntax is smart in that you don't have to list and name an attribute but rather rely on an implicit 'value' attribute.
+
+The current syntax, in which an underlying value type is referenced through 'oftype', it does not use instantiation/inheritance but rather composition.
+The purpose of using composition is that it is
+
+1. more similar to traditional approaches and
+2. is needed anyway for multi-attribute value types.
+
+Thus, this RFC removes the `oftype` keyword in a value type definitions and proposes an attribute-based syntax that is extendable to a future multi-attribute syntax.
+
+## Explanation
+
+I propose to make that attribute explicit. The new syntax would be:
 
 ```jayvee
 valuetype CorrelationCoefficient {
-  attributes: [
-    value oftype decimal;
-  ];
-  constraints: [
-    MinusOneToPlusOneRange; // Don't know how to attach this to value attribute
-  ];
+  property correlation oftype decimal;
+
+  constraint minusOneToPlusOneRange: MinusOneToPlusOneRange(value = correlation);
 }
+
+constraint MinusOneToPlusOneRange on decimal:
+  value >= -1 and value <= 1;
 ```
 
-While more verbose, it prepares the way for 
+The syntax is kept flat rather than using arrays.
+
+Despite referencing reusable constraints, the syntax also allows in-place constraint definition:
```

Comment:
I would keep this out of the RFC to make it smaller. We should probably have in-place constraint definitions but imho can do so in a follow-up.

Reply:
I wouldn't mind adding it here. It's not huge anyway.

Reply:
Just for a view on the syntax, if we go with my suggestion, that would be:

```
constraints: [ value >= 1 and value <=1 on correlation ]
```

equivalent to

```
constraints: [ MinusOneToPlusOneRange on correlation ]
```

My issue with that was that reading the in-place constraint definition with the following "on" is kind of confusing... But honestly, seeing how it looks here I am kind of happy with it. So I'll revert what I said and say we should have in place constraints as well, yeah.

Reply:
Why shouldn't we use
```
constraints: [ correlation >= 1 and correlation <=1 ]
```
?

And combined:
```
constraints: [
    correlation >= 1 and correlation <=1,
    MinusOneToPlusOneRange on correlation
]
```
I don't see why we need to use `on` for in-place constraints as well

Reply:
I use `correlation` instead of `value`.
This allows for example
```
constraints: [
    correlation >= 1 and otherUnit <=1
]
```

Reply:
Ah, fair enough. I dislike that because it introduces different syntax for inline constraints vs. references and I like the ability to always just create a constraint reference by c&p the inline code. But that is something that reasonable people can disagree on :D.

Reply:
I see your point @rhazn, and I see value in keeping the syntax consistent independently if defined in-line or as stand-alone element.

What do you think about changing the **constraint definition** syntax in a similar way to keep it consistent? Instead of the `value` keyword, constraint definitions would refer to the value type's properties they are based on:

```
constraint CorrelationCoefficient on decimal:
  value >= 1 and value <= 1; // "value" is the property name of builtin value types (so nothing changes for them)
  
constraint PostiveCorrelationConstraint on CorrelationCoefficient:
  correlation > 0; // "correlation" is a property of the value type CorrelationCoefficient
  
// Future, for multi-dimensional value types
constraint Coordinate2DInLeftUpperQuadrantConstraint on Coordinate2D:
  x > 0 and y > 0;
```

An obvious downside is the circular dependency of value type and constraint that could be incentivized (or should be prevented by hinting to in-line constraints in such cases).

Reply:
I rewrote this comment three times, not sure I can come to a strong opinion so in the end I would probably accept majority opinion, but:

I think that is unfortunate because it again introduces magic with the `value` name for builtin value types. It also still means it is weird to reuse constraints between inline definitions and general definitions. I would prefer if we keep `value` in general for one-dimensional value types.

I can not really think of a good way to model constrains on multi-dimensional value types that do not need to know the names of properties... For consistencies sake I would go with `value.x`. That does not resolve the issue of course, but it keeps the rule intact that `value` always refers to the underlying value of the value type.

Reply:
Ok, I see two alternatives:

## Alternative 1:  Inline constraints use property names

So in the example, `correlation >= -1 and correlation <= 1`?

This leads to different syntax for inline and referenceable constraints. 
However, users can call their property of single-attribute value types "value," and we can promote that through our documentation. Then, there is no syntactic difference when extracting an in-line constraint to a referenceable constraint.

```
// Referenceable constraint
valuetype CorrelationCoefficient {
  property correlation oftype decimal; 
  constraint minusOneToPlusOneRange: CorrelationCoefficient on correlation;
}
constraint CorrelationCoefficient on decimal:
  value >= -1 and value <= 1;

// Inline constraint (not following convention)
valuetype CorrelationCoefficient {
  property correlation oftype decimal; 

  constraint minusOneToPlusOneRange:
    correlation >= -1 and correlation <=1;
}

// Inline constraint (following convention)
valuetype CorrelationCoefficient {
  property value oftype decimal; // convention: should be called value for one-dim value types

  constraint minusOneToPlusOneRange:
    value >= -1 and value <=1;
}
```

## Alternative 2:  Constraints always use `value`

So in the example, `value.correlation >= -1 and value.correlation <= 1`. 
This alternative leads to similar syntax for inline and referenceable constraints.

```
// Referenceable constraint 
valuetype CorrelationCoefficient {
  property correlation oftype decimal; 
  constraint minusOneToPlusOneRange: CorrelationCoefficient on correlation;
}
constraint CorrelationCoefficient on decimal:
  value >= -1 and value <= 1;

// Inline constraint (not following convention)
valuetype CorrelationCoefficient {
  property correlation oftype decimal; 

  constraint minusOneToPlusOneRange on correlation:
    value >= -1 and value <=1;
}
```

The question remaining for this alternative  is how to model multi-dimensional constraints, like @rhazn already mentioned... I tried a few times, but didn't come to a good solution.